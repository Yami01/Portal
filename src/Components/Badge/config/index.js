/**
 *
 * @format
 * @flow
 *
 */

export const BADGE_SIZE = (size: number): number => {
  if (size > 0 && size) {
    return size
  } else {
    return 20
  }
}

export const BADGE_RADIUS = (size: number): number => {
  if (size > 0 && size) {
    return size / 2
  } else {
    return 10
  }
}
