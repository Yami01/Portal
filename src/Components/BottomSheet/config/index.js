/**
 *
 * @format
 * @flow
 *
 */
export const DEFAULT_BOTTOM_PADDING = 16
