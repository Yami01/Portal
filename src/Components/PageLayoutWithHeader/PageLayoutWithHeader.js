/**
 *
 * PageLayoutWithHeader
 * @format
 * @flow
 *
 */

import React, {
  memo,
  useMemo,
  useCallback,
  useState,
  useRef,
  useEffect,
} from 'react'
import type { Node, AbstractComponent } from 'react'
import { View, ScrollView, StatusBar } from 'react-native'
import Animated, {
  call,
  EasingNode,
  useCode,
  useValue as useReanimatedValue,
} from 'react-native-reanimated'
import type {
  LayoutEvent,
  ScrollEvent,
} from 'react-native/Libraries/Types/CoreEventTypes'

import {
  HEADER_TEST_ID,
  SCROLL_VIEW_TEST_ID,
  SHOW_HEADER_ANIMATION_DURATION,
} from './config'

import PageLayoutWithHeaderStyles from './styles'
import type { PropsType, UseCodeReturnType } from './types'
import Header from '@/Components/Header'

const PageLayoutWithHeader: AbstractComponent<PropsType> = memo(
  (props: PropsType): Node => {
    const {
      children,
      headerBackgroundColor,
      headerTitleLabel,
      headerTitleColor,
      headerTitleOnPress,
      headerTitleIconGroup,
      headerTitleIconName,
      headerTitleIconWidth,
      headerLeftButton,
      headerRightButton,
      isHeaderProgressBarVisible,
      headerProgressBarPosition,
      headerProgressBarBottomRightText,
      headerExtraStyle,
      isHeaderHidden,
      isHeaderAnimated = true,
      isHeaderAbsolute: isHeaderAbsoluteProp,
      isHeaderOnlyLeftIcon: isHeaderOnlyLeftIconProp,
      isHeaderOnlyRightIcon,
      isHeaderWithShadow,
      isHeaderAllButtonsOnRight,
      isHeaderTitleAlignLeft,
      isHeaderLargePadding,
      scrollYOffsetToSolidHeader,
      shouldHeaderIgnoreStatusBarHeight,
      headerCard,
      headerCardOverlapHeight = 0,
      scrollViewContentContainerStyle,
      scrollViewRef,
      scrollViewScrollEnabled,
      onScrollViewPositionChange,
      testID,
    } = props

    const prevScrollYRef = useRef(0)

    // Position. Larger means scrolling up
    const scrollYAnim = useReanimatedValue<number>(0)

    // Record Header height with extra bottom padding
    const totalHeaderHeightAnim = useReanimatedValue<number>(0)

    const isAnimatingHeaderMarginTopRef = useRef(false)
    const headerMarginTopAnim = useReanimatedValue<number>(0)

    const headerExtraPaddingAnim = useReanimatedValue<number>(
      headerCardOverlapHeight,
    )

    const [isShowDuplicatedHeader, setIsShowDuplicatedHeader] = useState(false)

    // Scroll Y for RN ScrollView
    const [scrollY, recordScrollY] = useState(0)

    // Update headerExtraPaddingAnim if headerCardOverlapHeight changed
    useEffect(() => {
      headerExtraPaddingAnim.setValue(headerCardOverlapHeight)
    }, [headerCardOverlapHeight, headerExtraPaddingAnim])

    // Control header position and extra padding height
    useCode((): UseCodeReturnType => {
      return call<number>(
        [scrollYAnim, headerMarginTopAnim, totalHeaderHeightAnim],
        ([
          scrollYAnimValue,
          headerMarginTop,
          totalHeaderHeight,
        ]: $ReadOnlyArray<number>) => {
          const prevScrollY = prevScrollYRef.current
          prevScrollYRef.current = scrollYAnimValue
          // +ve: scrolling up, -ve: scrolling down
          const scrollYDiff = scrollYAnimValue - prevScrollY
          // If scrollYAnimValue changed
          if (scrollYDiff) {
            // If scrolling down
            if (
              prevScrollY !== scrollYAnimValue &&
              prevScrollY > scrollYAnimValue
            ) {
              // Show Header if header not fully shown
              if (headerMarginTop < 0) {
                // If header marginTop animation not running
                if (!isAnimatingHeaderMarginTopRef.current) {
                  isAnimatingHeaderMarginTopRef.current = true
                  // Remove Header extra padding
                  setIsShowDuplicatedHeader(true)
                  headerExtraPaddingAnim.setValue(0)
                  // Animate header from top
                  Animated.timing(headerMarginTopAnim, {
                    toValue: 0,
                    duration: SHOW_HEADER_ANIMATION_DURATION,
                    easing: EasingNode.inOut(EasingNode.ease),
                  }).start(() => {
                    isAnimatingHeaderMarginTopRef.current = false
                  })
                }
              }
              // Increase Header extra padding when reach top
              if (scrollYAnimValue < headerCardOverlapHeight) {
                const newHeaderExtraPadding =
                  headerCardOverlapHeight - scrollYAnimValue
                headerExtraPaddingAnim.setValue(newHeaderExtraPadding)
                // Hide duplicated Header when full extra padding shown
                if (newHeaderExtraPadding >= headerCardOverlapHeight) {
                  setIsShowDuplicatedHeader(false)
                }
              }
            } else if (!isAnimatingHeaderMarginTopRef.current) {
              // Set Header position by scroll diff
              const newHeaderMarginTop = -Math.min(
                Math.max(0, -headerMarginTop + scrollYDiff),
                totalHeaderHeight,
              )
              headerMarginTopAnim.setValue(newHeaderMarginTop)
            }
          }
        },
      )
    }, [scrollYAnim, headerMarginTopAnim, totalHeaderHeightAnim])

    const onHeaderLayout = useCallback(
      (e: LayoutEvent) => {
        const {
          nativeEvent: {
            layout: { height },
          },
        } = e
        totalHeaderHeightAnim.setValue(height)
      },
      [totalHeaderHeightAnim],
    )

    const onScrollViewScroll = useCallback(
      (e: ScrollEvent) => {
        recordScrollY(e.nativeEvent.contentOffset.y)
        if (onScrollViewPositionChange) {
          onScrollViewPositionChange(e.nativeEvent.contentOffset.y)
        }
      },
      [onScrollViewPositionChange],
    )

    const isAllowHeaderChangeFromIconToSolid = useMemo((): boolean => {
      return !isHeaderAnimated && !!isHeaderOnlyLeftIconProp
    }, [isHeaderAnimated, isHeaderOnlyLeftIconProp])

    const isHeaderOnlyLeftIcon = useMemo((): boolean => {
      // If isHeaderAnimated true, always false
      // If isHeaderAnimated false and isHeaderOnlyLeftIcon false, use isHeaderOnlyLeftIcon (false)
      // If isHeaderAnimated false, isHeaderOnlyLeftIcon true and scrollYOffsetToSolidHeader provided, determine by scrollY
      // If isHeaderAnimated false, isHeaderOnlyLeftIcon true and scrollYOffsetToSolidHeader undefined, use isHeaderOnlyLeftIcon (true)
      if (isAllowHeaderChangeFromIconToSolid) {
        if (typeof scrollYOffsetToSolidHeader === 'number') {
          return scrollY <= (scrollYOffsetToSolidHeader || 0)
        }
        return isHeaderOnlyLeftIconProp || false
      }
      return false
    }, [
      isAllowHeaderChangeFromIconToSolid,
      isHeaderOnlyLeftIconProp,
      scrollY,
      scrollYOffsetToSolidHeader,
    ])

    const isHeaderAbsolute = useMemo((): boolean => {
      return (
        (isAllowHeaderChangeFromIconToSolid && isHeaderAbsoluteProp) || false
      )
    }, [isAllowHeaderChangeFromIconToSolid, isHeaderAbsoluteProp])

    const headerProps = useMemo((): HeaderPropsType => {
      return {
        isOnlyLeftIcon: isHeaderOnlyLeftIcon,
        isOnlyRightIcon: isHeaderOnlyRightIcon,
        backgroundColor: headerBackgroundColor,
        withShadow: isHeaderWithShadow,
        titleLabel: headerTitleLabel,
        titleColor: headerTitleColor,
        titleOnPress: headerTitleOnPress,
        titleIconGroup: headerTitleIconGroup,
        titleIconName: headerTitleIconName,
        titleIconWidth: headerTitleIconWidth,
        leftButton: headerLeftButton,
        rightButton: headerRightButton,
        isAllButtonsOnRight: isHeaderAllButtonsOnRight,
        isTitleAlignLeft: isHeaderTitleAlignLeft,
        isLargePadding: isHeaderLargePadding,
        shouldIgnoreStatusBarHeight: shouldHeaderIgnoreStatusBarHeight,
        isProgressBarVisible: isHeaderProgressBarVisible,
        progressBarPosition: headerProgressBarPosition,
        progressBarBottomRightText: headerProgressBarBottomRightText,
      }
    }, [
      headerBackgroundColor,
      headerLeftButton,
      headerRightButton,
      headerTitleColor,
      headerTitleIconGroup,
      headerTitleIconName,
      headerTitleIconWidth,
      headerTitleLabel,
      headerTitleOnPress,
      isHeaderAllButtonsOnRight,
      isHeaderLargePadding,
      isHeaderOnlyLeftIcon,
      isHeaderOnlyRightIcon,
      isHeaderTitleAlignLeft,
      isHeaderWithShadow,
      shouldHeaderIgnoreStatusBarHeight,
      isHeaderProgressBarVisible,
      headerProgressBarPosition,
      headerProgressBarBottomRightText,
    ])

    const headerNode = useMemo((): Node => {
      return (
        <Header
          {...headerProps}
          extraPaddingBottom={headerExtraPaddingAnim}
          extraStyle={headerExtraStyle}
          testID={HEADER_TEST_ID}
          onLayout={onHeaderLayout}
        />
      )
    }, [headerProps, headerExtraPaddingAnim, headerExtraStyle, onHeaderLayout])

    const duplicatedHeaderNode = useMemo((): Node => {
      return (
        isHeaderAnimated && (
          <Animated.View
            style={[
              PageLayoutWithHeaderStyles.duplicatedHeaderContainer,
              {
                marginTop: headerMarginTopAnim,
                opacity: isShowDuplicatedHeader ? 1 : 0,
              },
            ]}
            accessible={false}
            importantForAccessibility="no-hide-descendants"
            accessibilityElementsHidden
          >
            <Header {...headerProps} />
          </Animated.View>
        )
      )
    }, [
      headerMarginTopAnim,
      headerProps,
      isHeaderAnimated,
      isShowDuplicatedHeader,
    ])

    const isNormalHeaderShow = !isHeaderHidden && !isHeaderAbsolute
    const isAbsoluteHeaderShow = !isHeaderHidden && isHeaderAbsolute
    const ref = (scrollViewRef: any)

    return (
      <View style={PageLayoutWithHeaderStyles.container} {...testID}>
        {!shouldHeaderIgnoreStatusBarHeight && (
          <StatusBar backgroundColor={headerBackgroundColor} />
        )}
        {isHeaderAnimated ? (
          <>
            <Animated.ScrollView
              testID={SCROLL_VIEW_TEST_ID}
              scrollEventThrottle={16}
              onScroll={Animated.event(
                [{ nativeEvent: { contentOffset: { y: scrollYAnim } } }],
                {
                  useNativeDriver: true,
                },
              )}
              bounces={false}
            >
              {headerNode}

              <Animated.View
                style={[
                  PageLayoutWithHeaderStyles.scrollView,
                  {
                    marginTop: Animated.multiply(headerExtraPaddingAnim, -1),
                  },
                ]}
              >
                <View style={scrollViewContentContainerStyle}>
                  {headerCard}
                  {children}
                </View>
              </Animated.View>
            </Animated.ScrollView>

            {duplicatedHeaderNode}
          </>
        ) : (
          <>
            <ScrollView
              contentContainerStyle={scrollViewContentContainerStyle}
              testID={SCROLL_VIEW_TEST_ID}
              stickyHeaderIndices={isHeaderAbsolute ? undefined : [0]}
              bounces={false}
              scrollEventThrottle={16}
              onScroll={onScrollViewScroll}
              ref={ref}
              scrollEnabled={scrollViewScrollEnabled}
            >
              {isNormalHeaderShow && headerNode}
              {children}
            </ScrollView>
            {isAbsoluteHeaderShow && (
              <View style={PageLayoutWithHeaderStyles.headerAbsoluteContainer}>
                {headerNode}
              </View>
            )}
          </>
        )}
      </View>
    )
  },
)

export default PageLayoutWithHeader
