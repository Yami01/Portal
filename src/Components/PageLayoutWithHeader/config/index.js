/**
 *
 * @format
 * @flow
 *
 */

export const HEADER_TEST_ID = 'headerTestId'
export const SCROLL_VIEW_TEST_ID = 'scrollViewTestId'

export const SHOW_HEADER_ANIMATION_DURATION = 300
