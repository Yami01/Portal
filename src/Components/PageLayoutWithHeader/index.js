/**
 *
 * @flow
 *
 */

export { default } from './PageLayoutWithHeader'
