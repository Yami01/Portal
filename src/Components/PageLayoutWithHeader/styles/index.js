/**
 *
 * @format
 * @flow
 *
 */

import { StyleSheet } from 'react-native'

import type { PageLayoutWithHeaderStylesType } from '../types'

const PageLayoutWithHeaderStyles: PageLayoutWithHeaderStylesType =
  StyleSheet.create({
    container: {
      flex: 1,
    },
    duplicatedHeaderContainer: {
      position: 'absolute',
      top: 0,
      width: '100%',
    },
    headerAbsoluteContainer: {
      position: 'absolute',
      top: 0,
      width: '100%',
    },
    scrollView: {
      zIndex: 99,
      elevation: 99,
    },
  })

export default PageLayoutWithHeaderStyles
