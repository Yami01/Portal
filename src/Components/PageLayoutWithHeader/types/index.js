/**
 *
 * @format
 * @flow
 *
 */
import type { Node } from 'react'
import type { ScrollViewImperativeMethods } from 'react-native/Libraries/Components/ScrollView/ScrollView'
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet'
import type { PressEvent } from 'react-native/Libraries/Types/CoreEventTypes'
import type { AnimatedNode } from 'react-native-reanimated'

export type HeaderButtonConfigType = {
  iconGroup: string,
  iconName: string,
  iconFillColor: string,
  // Only button right can have label and badgeNumber
  label?: string,
  badgeNumber?: number,
  isShowDot?: boolean,
  testID?: string,
  onPress?: (event: PressEvent) => any,
}

export type PropsType = {
  headerBackgroundColor?: string,
  headerTitleLabel?: string,
  headerTitleColor?: string,
  headerTitleOnPress?: () => void,
  headerTitleIconGroup?: string,
  headerTitleIconName?: string,
  headerTitleIconWidth?: number,
  headerLeftButton?: HeaderButtonConfigType,
  headerRightButton?: HeaderButtonConfigType,
  isHeaderProgressBarVisible?: boolean,
  headerProgressBarPosition?: number,
  headerProgressBarBottomRightText?: string,
  headerExtraStyle?: ViewStyleProp,
  isHeaderHidden?: boolean,
  isHeaderAnimated?: boolean,
  isHeaderOnlyLeftIcon?: boolean,
  isHeaderOnlyRightIcon?: boolean,
  isHeaderAbsolute?: boolean,
  isHeaderWithShadow?: boolean,
  isHeaderAllButtonsOnRight?: boolean,
  isHeaderTitleAlignLeft?: boolean,
  isHeaderLargePadding?: boolean,
  scrollYOffsetToSolidHeader?: number,
  shouldHeaderIgnoreStatusBarHeight?: boolean,
  headerCard?: Node,
  headerCardOverlapHeight?: number,
  children?: Node,
  scrollViewContentContainerStyle?: ViewStyleProp,
  scrollViewRef?: { current: ?ScrollViewImperativeMethods },
  scrollViewScrollEnabled?: boolean,
  onScrollViewPositionChange?: (position: number) => void,
  testID?: string,
}

export type PageLayoutWithHeaderStylesType = {
  container: ViewStyleProp,
  duplicatedHeaderContainer: ViewStyleProp,
  headerAbsoluteContainer: ViewStyleProp,
  scrollView: ViewStyleProp,
}

const zero = 0
export type UseCodeReturnType = AnimatedNode<typeof zero>
