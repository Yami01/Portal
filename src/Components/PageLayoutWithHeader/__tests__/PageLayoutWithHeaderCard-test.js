/**
 *
 * PageLayoutWithHeader Test
 * @format
 * @flow
 *
 */

import React from 'react';
import renderer from 'react-test-renderer';
import {render, within, fireEvent, act} from '@testing-library/react-native';

import {View, Text} from 'react-native';

import {WHITE} from '@colors';
import PageLayoutWithHeader from '../PageLayoutWithHeader';
import {HEADER_TEST_ID, SCROLL_VIEW_TEST_ID} from '../config';

const props = {
  headerTitleLabel: 'Header',
  headerAccessibilityLabel: 'Accessible Header',
  headerAccessibilityHint: 'This is accessible header',
  headerCardOverlapHeight: 30,
};

jest.useFakeTimers();

describe('PageLayoutWithHeader', () => {
  const leftButtonAccessibilityLabel = 'Accessible header button left';
  const leftButton = {
    iconFillColor: WHITE,
    iconGroup: 'general',
    iconName: 'menu',
    accessibilityLabel: leftButtonAccessibilityLabel,
    accessibilityHint: 'This button will take you home',
  };

  const rightButtonAccessibilityLabel = 'Accessible header button right';
  const rightButton = {
    iconFillColor: WHITE,
    iconGroup: 'general',
    iconName: 'notifications',
    accessibilityLabel: rightButtonAccessibilityLabel,
    accessibilityHint: 'This button will take to the notification',
    badgeNumber: 1,
  };

  const titleLabel = 'AIA Vitality';

  const headerProps = {
    isHeaderProgressBarVisible: true,
    headerProgressBarPosition: 0,
    headerProgressBarAccessibilityLabel: 'progressBarAccessibilityLabel',
    headerProgressBarAccessibilityHint: 'progressBarAccessibilityHint',
    headerProgressBarBottomRightText: '0% completed',
    headerProgressBarBottomRightTextAccessibilityLabel:
      'progressBarBottomRightTextAccessibilityLabel',
    headerProgressBarBottomRightTextAccessibilityHint:
      'progressBarBottomRightTextAccessibilityHint',
  };

  const getScrollEventData = (scrollY: number): Obj => {
    return {
      nativeEvent: {
        contentOffset: {
          y: scrollY,
        },
      },
    };
  };

  test('render with animated header properly', () => {
    const tree = renderer
      .create(
        <PageLayoutWithHeader
          {...props}
          headerCard={
            <View>
              <Text>Card</Text>
            </View>
          }>
          <View>
            <Text>Children</Text>
          </View>
        </PageLayoutWithHeader>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('render with non animated header properly', () => {
    const tree = renderer
      .create(
        <PageLayoutWithHeader
          {...props}
          headerLeftButton={leftButton}
          headerTitleLabel={titleLabel}
          headerRightButton={rightButton}
          isHeaderAnimated={false}
          isHeaderOnlyLeftIcon
          isHeaderAbsolute
          scrollYOffsetToSolidHeader={100}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('is render properly', () => {
    const cardTestId = 'Card Test Id';
    const childrenTestId = 'Children Test Id';
    const {queryByTestId} = render(
      <PageLayoutWithHeader {...props} headerCard={<View testID={cardTestId} />}>
        <View testID={childrenTestId} />
      </PageLayoutWithHeader>
    );
    expect(queryByTestId(cardTestId)).not.toBeNull();
    expect(queryByTestId(childrenTestId)).not.toBeNull();
  });

  test('is isHeaderOnlyLeftIcon not works when isHeaderAnimated true', () => {
    const {queryAllByA11yLabel, queryAllByText} = render(
      <PageLayoutWithHeader
        {...props}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated
        isHeaderOnlyLeftIcon
      />
    );
    expect(queryAllByA11yLabel(leftButtonAccessibilityLabel)).not.toBeNull();
    expect(queryAllByText(titleLabel)).not.toBeNull();
    expect(queryAllByA11yLabel(rightButtonAccessibilityLabel)).not.toBeNull();
  });

  test('is isHeaderOnlyLeftIcon work when isHeaderAnimated false', () => {
    const {queryByA11yLabel, queryByText} = render(
      <PageLayoutWithHeader
        {...props}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated={false}
        isHeaderOnlyLeftIcon
      />
    );
    expect(queryByA11yLabel(leftButtonAccessibilityLabel)).not.toBeNull();
    expect(queryByText(titleLabel)).toBeNull();
    expect(queryByA11yLabel(rightButtonAccessibilityLabel)).toBeNull();
  });

  test('is Header is absolute position when isHeaderAbsolute true', () => {
    const {queryByTestId, queryAllByA11yLabel} = render(
      <PageLayoutWithHeader
        {...props}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated={false}
        isHeaderOnlyLeftIcon
        isHeaderAbsolute
      />
    );
    const scrollView = queryByTestId(SCROLL_VIEW_TEST_ID);
    expect(scrollView).not.toBeNull();

    const headerLeftButtons = queryAllByA11yLabel(leftButtonAccessibilityLabel);
    expect(headerLeftButtons).not.toBeNull();
    expect(headerLeftButtons).toHaveLength(1);

    // If left button outside scroll view
    const {queryByA11yLabel: queryByA11yLabelInScrollView} = within(scrollView);
    expect(queryByA11yLabelInScrollView(leftButtonAccessibilityLabel)).toBeNull();
  });

  test('is Header is relative position when isHeaderAbsolute false', () => {
    const {queryByTestId, queryAllByA11yLabel} = render(
      <PageLayoutWithHeader
        {...props}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated={false}
        isHeaderOnlyLeftIcon
        isHeaderAbsolute={false}
      />
    );
    const scrollView = queryByTestId(SCROLL_VIEW_TEST_ID);
    expect(scrollView).not.toBeNull();

    const headerLeftButtons = queryAllByA11yLabel(leftButtonAccessibilityLabel);
    expect(headerLeftButtons).not.toBeNull();
    expect(headerLeftButtons).toHaveLength(1);

    // If left button inside scroll view
    const {queryByA11yLabel: queryByA11yLabelInScrollView} = within(scrollView);
    expect(queryByA11yLabelInScrollView(leftButtonAccessibilityLabel)).not.toBeNull();
  });

  test('is headerCard not show when isHeaderAnimated false', () => {
    const cardTestId = 'Card Test Id';
    const childrenTestId = 'Children Test Id';
    const {queryByTestId} = render(
      <PageLayoutWithHeader
        {...props}
        headerCard={<View testID={cardTestId} />}
        isHeaderAnimated={false}>
        <View testID={childrenTestId} />
      </PageLayoutWithHeader>
    );
    expect(queryByTestId(cardTestId)).toBeNull();
    expect(queryByTestId(childrenTestId)).not.toBeNull();
  });

  test('is ScrollView onScroll working for non animated header', () => {
    const scrollYOffset = 50;
    const {queryByTestId} = render(
      <PageLayoutWithHeader
        {...props}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated={false}
        isHeaderOnlyLeftIcon
        scrollYOffsetToSolidHeader={scrollYOffset}>
        <View style={{height: 1000}} />
      </PageLayoutWithHeader>
    );
    const scrollView = queryByTestId(SCROLL_VIEW_TEST_ID);
    expect(scrollView).not.toBeNull();

    fireEvent.scroll(scrollView, getScrollEventData(scrollYOffset + 1));
  });

  test('is Header onLayout working', () => {
    const {queryByTestId} = render(<PageLayoutWithHeader {...props} />);
    const header = queryByTestId(HEADER_TEST_ID);
    expect(header).not.toBeNull();
    act(() => {
      fireEvent(header, 'onLayout', {
        nativeEvent: {layout: {height: 120}},
      });
    });
  });

  test('is Progress Bar work', () => {
    const {queryByA11yLabel, queryByText, queryByA11yHint} = render(
      <PageLayoutWithHeader
        {...props}
        {...headerProps}
        headerLeftButton={leftButton}
        headerTitleLabel={titleLabel}
        headerRightButton={rightButton}
        isHeaderAnimated={false}
        isHeaderOnlyLeftIcon
      />
    );
    expect(queryByA11yLabel(headerProps.headerProgressBarAccessibilityLabel)).not.toBeNull();
    expect(queryByA11yHint(headerProps.headerProgressBarAccessibilityHint)).not.toBeNull();
    expect(queryByText(headerProps.headerProgressBarBottomRightText)).not.toBeNull();
    expect(
      queryByA11yLabel(headerProps.headerProgressBarBottomRightTextAccessibilityLabel)
    ).not.toBeNull();
    expect(
      queryByA11yHint(headerProps.headerProgressBarBottomRightTextAccessibilityHint)
    ).not.toBeNull();
  });
});
