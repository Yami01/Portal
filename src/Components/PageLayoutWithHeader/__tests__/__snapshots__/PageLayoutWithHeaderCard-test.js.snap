// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLayoutWithHeader render with animated header properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    scrollEventThrottle={16}
    style={Object {}}
    testID="scrollViewTestId"
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#d31145",
              "borderBottomLeftRadius": 16,
              "flexDirection": "row",
              "overflow": "visible",
              "paddingHorizontal": 16,
              "width": "100%",
              "zIndex": 98,
            },
            Object {
              "backgroundColor": "#d31145",
              "paddingBottom": AnimatedValue {
                " __value": 46,
              },
              "paddingTop": 36,
            },
            false,
            undefined,
          ]
        }
        testID="headerTestId"
      >
        <View
          accessibilityHint="This is accessible header"
          accessibilityLabel="Accessible Header"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                Object {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <View
              accessibilityRole="text"
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  false,
                  Object {
                    "alignItems": "center",
                    "alignSelf": "flex-start",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                ]
              }
              testID="titleIconTextTestID"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "alignSelf": "flex-start",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      undefined,
                      Array [
                        Object {
                          "fontSize": 17,
                          "lineHeight": 24,
                        },
                        Object {
                          "color": "#fff",
                        },
                      ],
                    ]
                  }
                >
                  Header
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "elevation": 99,
              "zIndex": 99,
            },
            Object {
              "marginTop": AnimatedValue {
                " __value": -30,
              },
            },
          ]
        }
      >
        <View>
          <View>
            <Text>
              Card
            </Text>
          </View>
          <View>
            <Text>
              Children
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityElementsHidden={true}
    accessible={false}
    importantForAccessibility="no-hide-descendants"
    style={
      Array [
        Object {
          "position": "absolute",
          "top": 0,
          "width": "100%",
        },
        Object {
          "marginTop": AnimatedValue {
            " __value": 0,
          },
          "opacity": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#d31145",
            "borderBottomLeftRadius": 16,
            "flexDirection": "row",
            "overflow": "visible",
            "paddingHorizontal": 16,
            "width": "100%",
            "zIndex": 98,
          },
          Object {
            "backgroundColor": "#d31145",
            "paddingBottom": AnimatedValue {
              " __value": 16,
            },
            "paddingTop": 36,
          },
          false,
          undefined,
        ]
      }
    >
      <View
        accessibilityHint="This is accessible header"
        accessibilityLabel="Accessible Header"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              Object {
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            accessibilityRole="text"
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                false,
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
              ]
            }
            testID="titleIconTextTestID"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Array [
                    undefined,
                    Array [
                      Object {
                        "fontSize": 17,
                        "lineHeight": 24,
                      },
                      Object {
                        "color": "#fff",
                      },
                    ],
                  ]
                }
              >
                Header
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`PageLayoutWithHeader render with non animated header properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    onScroll={[Function]}
    scrollEventThrottle={16}
    testID="scrollViewTestId"
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#d31145",
            "borderBottomLeftRadius": 16,
            "flexDirection": "row",
            "overflow": "visible",
            "paddingHorizontal": 16,
            "width": "100%",
            "zIndex": 98,
          },
          Object {
            "backgroundColor": "rgba(0,0,0,0)",
            "paddingBottom": AnimatedValue {
              " __value": 46,
            },
            "paddingTop": 36,
          },
          false,
          undefined,
        ]
      }
      testID="headerTestId"
    >
      <View
        accessibilityHint="This is accessible header"
        accessibilityLabel="Accessible Header"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 24,
                "minWidth": 24,
              },
              Object {
                "marginRight": 12,
              },
            ]
          }
          testID="leftButtonContainerTestID"
        >
          <View>
            <View
              accessibilityHint="This button will take you home"
              accessibilityLabel="Accessible header button left"
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  false,
                  undefined,
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    },
                    Array [
                      Object {
                        "borderRadius": 20,
                        "elevation": 5,
                        "margin": -8,
                        "padding": 8,
                        "shadowColor": "#596c80",
                        "shadowOffset": Object {
                          "height": 3,
                          "width": 2,
                        },
                        "shadowOpacity": 0.15,
                        "shadowRadius": 2,
                      },
                      Object {
                        "backgroundColor": "#d31145",
                      },
                    ],
                  ]
                }
                testID="touchableWrapperTestID"
              >
                <View
                  style={
                    Array [
                      undefined,
                      undefined,
                    ]
                  }
                >
                  <svg
                    data-file-name="SvgMenu"
                    fill="#fff"
                    height={24}
                    width={24}
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
